Gen. 9 Move Effects, 1st batch

LOuroboros opened the pull request:
## Description
After a bit of a talk in the Discord server about how big PRs such as #2470 are a chore to review, it was seemingly agreed that we should try to limit the scope of the PRs a little to make them easier to diggest and review.
Thus, here's the 1st batch of Gen. 9 move effects out of 5.

Effects that won't be included in any of these PRs:
- Silk Trap, because Ghoulslash already did it :P
- Tera Blast, because its effect involves Terastalization, so I feel it's better to wait before adding it. At least until #2372 is merged.
- Order Up, because it's directly tied to Tatsugiri, Dondozo and the Commander ability which has yet to be implemented.
  - It also relies on Tatsugiri's forms, which just like the species itself, are not implemented.
- Revival Blessing, because I'm not confident I can make a nice and clean implementation.
- Raging Bull, because it's directly tied to Paldean Tauros and its forms which have yet to be implemented.
- Chilly Reception, because it requires the Snow battle weather which I couldn't add successfully without the game either freezing or glitching out.

Closes #2479, #2483, #2478, #2479 and #2480 btw.

## **Discord contact info**
Lunos#4026

mrgriffin commented on src/battle_script_commands.c:
```diff
@@ -3647,6 +3647,22 @@ void SetMoveEffect(bool32 primary, u32 certain)
+                else
+                {
+                DIRE_CLAW_STATUS_PICK:
+                    gBattleScripting.moveEffect = Random() % 6;
+                    if (gBattleScripting.moveEffect == 0
+                     || gBattleScripting.moveEffect == MOVE_EFFECT_BURN
+                     || gBattleScripting.moveEffect == MOVE_EFFECT_FREEZE)
+                        goto DIRE_CLAW_STATUS_PICK;
```
I'd potentially write this more like:
```c
static const u8 sDireClawStatuses[] = { MOVE_EFFECT_POISON, MOVE_EFFECT_PARALYSIS, MOVE_EFFECT_SLEEP };

gBattleScripting.moveEffect = sDireClawStatuses[Random() % 3];
```

LOuroboros replied:
Ohh! :eyes:

I did want to use an array at first, but that wasn't working too well because back then I was trying to apply the `STATUS1` directly.

I'll test that out, thanks!

LOuroboros replied:
Tested it out after fighting with Shell Trap this entire week. Sadly, it doesn't work 😔

mrgriffin reviewed (commented):
I think your Shell Trap might be functioning incorrectly, Bulbapedia says this (emphasis mine):
> The user sets up a trap at the beginning of the turn, then **if it gets hit by an opponent's [physical move](https://bulbapedia.bulbagarden.net/wiki/Physical_move) within the same turn, it will use Shell Trap immediately after that move**. Otherwise, when the user's scheduled turn happens at -3 priority, Shell Trap will fail at that time.

I think your script makes the move go off when we get to our turn. I guess this is pretty close to being the same in singles, but quite different in doubles?

Bulbapedia also says:
> Shell Trap will not be triggered ... being hit by a move that's receiving a bonus from [Sheer Force](https://bulbapedia.bulbagarden.net/wiki/Sheer_Force_(Ability)).

So to me that suggests that Shell Trap's attack should trigger during move end, similar to Life Orb?

LOuroboros commented:
> The user sets up a trap at the beginning of the turn, then **if it gets hit by an opponent's [physical move](https://bulbapedia.bulbagarden.net/wiki/Physical_move) within the same turn, it will use Shell Trap immediately after that move**. Otherwise, when the user's scheduled turn happens at -3 priority, Shell Trap will fail at that time.

That's exactly what's happening right now :eyes:

![mGBA_20221222_111727001](https://user-images.githubusercontent.com/4485172/209153932-6fcbed55-ac30-4c49-9f11-61b5812c536b.gif)

> I think your script makes the move go off when we get to our turn. I guess this is pretty close to being the same in singles, but quite different in doubles?

Is it? How so? :eyes:

> Bulbapedia also says:

> >    Shell Trap will not be triggered ... being hit by a move that's receiving a bonus from [Sheer Force](https://bulbapedia.bulbagarden.net/wiki/Sheer_Force_(Ability)).

> So to me that suggests that Shell Trap's attack should trigger during move end, similar to Life Orb?

That could be the case. The way in which the checks are written in Bulbapedia does sound like this would fit better as a `MOVEEND` effect. I'll look into it.

LOuroboros commented:
Alrighty, so:

I removed Shell Trap. It's a Gen. 7 move which has no place in this PR, and I also couldn't get it to work exactly as it should, i.e. as it does in the official games.

I updated Dire Claw's effect's code following MGriffin's suggestion which looks much nicer.

I'd say this is ready for another review :eyes:

LOuroboros commented:
Closes #2479, #2483, #2478, #2479 and #2480 btw.

ghoulslash commented on data/battle_scripts_1.s:
```diff
@@ -417,13 +417,35 @@ gBattleScriptsForMoveEffects::
+	.4byte BattleScript_EffectBarbBarrage             @ EFFECT_BARB_BARRAGE
+
+BattleScript_EffectStoneAxe::
+	call BattleScript_EffectHit_Ret
```
I don't like having a copy of EffectHit with a return. I would suggest making a MOVE_EFFECT_STEALTH_ROCK that can be used here. In addition, it shouldnt set stealth rock if the battle will end (there are similar checks in other MOVE_EFFECT cases)

LOuroboros replied:
> I don't like having a copy of EffectHit with a return.

Why not? It's far, far better than having what is basically 95% of it copy-pasted :/

I already made use of it for some of the other effects in the upcoming batches, but that aside, I did it *because* it would make situations like this where you have move effects that are mostly copy-pastes of `EFFECT_HIT`'s own battle script and only deviate from it a little by the end easier to handle.

> I would suggest making a MOVE_EFFECT_STEALTH_ROCK that can be used here.

I mean, that sounds fine too, I guess? But is there an actual reason to do that?

> In addition, it shouldnt set stealth rock if the battle will end (there are similar checks in other MOVE_EFFECT cases)

Actually, it should. #JustGameFreakThings

https://user-images.githubusercontent.com/4485172/210582829-5f51a9a8-b274-4c1b-8f7f-21a4900107b9.mp4

ghoulslash commented on data/battle_scripts_1.s:
```diff
@@ -417,13 +417,35 @@ gBattleScriptsForMoveEffects::
+	goto BattleScript_EffectHit
+
+BattleScript_EffectCeaselessEdge::
+	call BattleScript_EffectHit_Ret
```
Same comment as above

ghoulslash reviewed (changes requested):
Additional AI checks are needed for the new effects

LOuroboros commented:
> I'll look into it.

Alright, done I think. That wasn't too scary :eyes:

LOuroboros commented:
Ready for another review :eyes:

SBird1337 commented:
w.r.t. #2564 merged I would like to see integration tests for the new moves. I'll volunteer to

1) assist in writing tests if someone needs help.
2) write some tests, if required.

LOuroboros commented:
Solved the merge conflicts :eyes:

LOuroboros commented:
Conflicts and syntax issues solved, ready for re-review :eyes:

DizzyEggg commented:
Do you have plans to write tests for new move effects? 👀

LOuroboros commented:
> Do you have plans to write tests for new move effects? 👀

Nope. I think SBird was interested in doing it though?

DizzyEggg commented on data/battle_scripts_1.s:
```diff
@@ -417,7 +417,77 @@ gBattleScriptsForMoveEffects::
+	.4byte BattleScript_EffectStoneAxe                @ EFFECT_STONE_AXE
+	.4byte BattleScript_EffectBarbBarrage             @ EFFECT_BARB_BARRAGE
+
+BattleScript_EffectStoneAxe::
```
Perhaps it would be a good idea to merge BattleScript_EffectCeaselessEdge and BattleScript_EffectStoneAxe into one battlescript? We could use the argument field in the battle moves to set which hazard to set up.

LOuroboros replied:
Done. Will push when I have every correction-suggestion ready :eyes:

DizzyEggg commented on src/battle_util.c:
```diff
@@ -8761,6 +8761,10 @@ static u16 CalcMoveBasePower(u16 move, u8 battlerAtk, u8 battlerDef)
         basePower = CalcBeatUpPower();
         #endif
         break;
+    case EFFECT_BARB_BARRAGE:
```
This could go to Venoshock's Effect

LOuroboros commented on data/battle_scripts_1.s:
```diff
@@ -417,7 +417,77 @@ gBattleScriptsForMoveEffects::
+	moveendall
+	end
+
+BattleScript_StealthRockActivates::
```
I'm not too confident about this because working with string tables in battle scripts has always been a gigantic pain for me, but I'll try. Wish me luck.

LOuroboros commented:
Done. I won't feel good if I don't express that I don't like one bit string tables. Or at least this one. It only makes the relevant code that uses it harder to read and work with for no reason as far as I can think.

In any case, it's ready for re-re-review.

DizzyEggg commented on data/battle_scripts_1.s:
```diff
@@ -417,7 +417,51 @@ gBattleScriptsForMoveEffects::
+	setmoveeffect MOVE_EFFECT_DIRE_CLAW
+	goto BattleScript_EffectHit
+
+BattleScript_EffectHitSetEntryHazard::
```
I think this could be simplified to
BattleScript_EffectHitSetEntryHazard::
argumenttomoveeffect
goto BattleScript_EffectHit

The effect chance in the move table would need to be changed to 100% however.

DizzyEggg commented on data/battle_scripts_1.s:
```diff
@@ -417,7 +417,51 @@ gBattleScriptsForMoveEffects::
+	moveendall
+	end
+
+BattleScript_SpikesActivate::
```
I think this needs a second BattleScript for setstealthrock 
sth like
BattleScript_StealthRockActivates::
setstealthrock BattleScript_MoveEnd
goto BattleScript_PrintHazards

LOuroboros replied:
?

That's what `BattleScript_StealthRockActivates` is though?

DizzyEggg commented on include/battle_scripts.h:
```diff
@@ -456,6 +456,8 @@ extern const u8 BattleScript_HealingWishActivates[];
 extern const u8 BattleScript_ShellTrapSetUp[];
 extern const u8 BattleScript_CouldntFullyProtect[];
+extern const u8 BattleScript_StealthRockActivates[];
+extern const u8 BattleScript_SpikesActivate[];
```
BattleScript_SpikesActivates

LOuroboros replied:
Why? "The Spikes [as in they] *activate*" 🤔 ...

LOuroboros replied:
I guess you could also say that "the side status known as Spikes activates"...?

Sorry, I just didn't think about it like that lol. I'll make the change.

I already made the changes to `BattleScript_EffectHitSetEntryHazard` and the `case MOVE_EFFECT_DIRE_CLAW` of `Cmd_setmoveeffect` fwiw. I was just waiting for your response in the comments for `BattleScript_SpikesActivate` and `activateAfterFaint`.

DizzyEggg commented on src/battle_script_commands.c:
```diff
@@ -2934,6 +2939,7 @@ void SetMoveEffect(bool32 primary, u32 certain)
         INCREMENT_RESET_RETURN
 
     if (gBattleMons[gEffectBattler].hp == 0
+        && !activateAfterFaint
```
Is the variable needed? You could just do the same as with the already existing move effects here.

gBattleScripting.moveEffect != MOVE_EFFECT_STEALTH_ROCK && gBattleScripting.moveEffect != MOVE_EFFECT_SPIKES

LOuroboros replied:
How so?

The variable is used to tell the code that the effects should activate even if the opponent faints.

I added it because they weren't.

LOuroboros replied:
Alright, I see what you mean now. Sorry, I didn't even bother to read the code earlier in the morning.

Though wouldn't it be better to keep the variable and instead, incorporate `MOVE_EFFECT_PAYDAY`, `MOVE_EFFECT_STEAL_ITEM` and `MOVE_EFFECT_BUG_BITE` into the `switch (gBattleScripting.moveEffect)`? 🤔 

Kinda feels things would be more readable like that.

E.g.
```diff
diff --git a/src/battle_script_commands.c b/src/battle_script_commands.c
index d0de44b72..336655b0e 100644
--- a/src/battle_script_commands.c
+++ b/src/battle_script_commands.c
@@ -2896,6 +2896,9 @@ void SetMoveEffect(bool32 primary, u32 certain)
         return;
     case MOVE_EFFECT_STEALTH_ROCK:
     case MOVE_EFFECT_SPIKES:
+    case MOVE_EFFECT_PAYDAY:
+    case MOVE_EFFECT_STEAL_ITEM:
+    case MOVE_EFFECT_BUG_BITE:
         activateAfterFaint = TRUE;
         break;
     }
@@ -2938,11 +2941,7 @@ void SetMoveEffect(bool32 primary, u32 certain)
     if (TestSheerForceFlag(gBattlerAttacker, gCurrentMove) && affectsUser != MOVE_EFFECT_AFFECTS_USER)
         INCREMENT_RESET_RETURN
 
-    if (gBattleMons[gEffectBattler].hp == 0
-        && !activateAfterFaint
-        && gBattleScripting.moveEffect != MOVE_EFFECT_PAYDAY
-        && gBattleScripting.moveEffect != MOVE_EFFECT_STEAL_ITEM
-        && gBattleScripting.moveEffect != MOVE_EFFECT_BUG_BITE)
+    if (gBattleMons[gEffectBattler].hp == 0 && !activateAfterFaint)
         INCREMENT_RESET_RETURN
 
     if (DoesSubstituteBlockMove(gBattlerAttacker, gEffectBattler, gCurrentMove) && affectsUser != MOVE_EFFECT_AFFECTS_USER)
```

LOuroboros replied:
I could rename `activateAfterFaint` to `activateAfterTargetFaints` for extra readability too.

LOuroboros replied:
Pinging @DizzyEggg to get a response 😅 ...

DizzyEggg commented on src/battle_script_commands.c:
```diff
@@ -3751,6 +3757,30 @@ void SetMoveEffect(bool32 primary, u32 certain)
+                if (!gBattleMons[gEffectBattler].status1)
+                {
+                    static const u8 sDireClawEffects[] = { MOVE_EFFECT_POISON, MOVE_EFFECT_PARALYSIS, MOVE_EFFECT_SLEEP };
+                    gBattleScripting.moveEffect = sDireClawEffects[Random() % 3];
```
Random() % ARRAY_COUNT(sDireClawEffects)

DizzyEggg commented:
Please fix conflicts

LOuroboros commented:
> Please fix conflicts

Alright, but you still haven't addressed my comments.

LOuroboros commented:
Done.

I also uh, kinda pushed the commit that I didn't want to push until I got your response, @DizzyEggg.
